opt: reduce unnecessary allocations in the optimizer

DrewKimball opened the pull request:
**opt: avoid unnecessary allocations when enforcing ordering**

The optimizer attempts to optimize and cost every expression without
requiring any ordering by adding a `SortExpr` as the expression's parent.
This can be called many times for each expression in the memo group, since
there are often many possible required properties for each group. In
sufficiently complex queries, allocating a `SortExpr` each time can cause
non-neglibile overhead. The expression with ordering enforced by a `SortExpr`
is often not even added to the memo, since this only happens if the result
has a lower cost than the alternative.

This commit adds a scratch `SortExpr` to the optimizer to be used when
optimizing an expression with the ordering enforced. The scratch `SortExpr`
is only reallocated if it gets added to the memo. This removes the
unnecessary allocations.

Release note: None

**opt: unexport relational stats to avoid accidental copies**

Previously, the `props.Statistics` field of `props.Relational` was unexported.
This made it easy to accidentally copy it, which could show up on heap profiles
because `props.Statistics` is a fairly large struct.

This patch unexports `props.Statistics` and exposes a `Relational.Statistics()`
method that returns a pointer to the field. This changes the default behavior
to reference the struct, rather than copying it, in order to prevent accidental
copies in the future (and revert the current ones).

Release note: None

**opt: reduce allocations when building join multiplicity**

This commit adds a `ComputeEquivClosureNoCopy` method to `FuncDepSet`
that performs the function of `ComputeEquivClosure` in-place on the given
`ColSet`. This allows callers to avoid an unnecessary copy if it is ok
to mutate the starting set.

This commit also modifies the multiplicity builder logic to simply build
the equivalence closure in one pass over the filters, rather than first
constructing a `FuncDepSet`. This significantly decreases allocations in
queries with many joins and equality filters.

Release note: None

DrewKimball committed:
opt: avoid unnecessary allocations when enforcing ordering

The optimizer attempts to optimize and cost every expression without
requiring any ordering by adding a `SortExpr` as the expression's parent.
This can be called many times for each expression in the memo group, since
there are often many possible required properties for each group. In
sufficiently complex queries, allocating a `SortExpr` each time can cause
non-neglibile overhead. The expression with ordering enforced by a `SortExpr`
is often not even added to the memo, since this only happens if the result
has a lower cost than the alternative.

This commit adds a scratch `SortExpr` to the optimizer to be used when
optimizing an expression with the ordering enforced. The scratch `SortExpr`
is only reallocated if it gets added to the memo. This removes the
unnecessary allocations.

Release note: None

DrewKimball committed:
opt: unexport relational stats to avoid accidental copies

Previously, the `props.Statistics` field of `props.Relational` was exported.
This made it easy to accidentally copy it, which could show up on heap profiles
because `props.Statistics` is a fairly large struct.

This patch unexports `props.Statistics` and exposes a `Relational.Statistics()`
method that returns a pointer to the field. This changes the default behavior
to reference the struct, rather than copying it, in order to prevent accidental
copies in the future (and revert the current ones).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87920)
<!-- Reviewable:end -->

DrewKimball commented:
Slow-queries benchmark vs master:
```
name                         old time/op    new time/op    delta
SlowQueries/slow-query-1-10    19.2ms ± 0%    18.8ms ± 1%   -2.01%  (p=0.000 n=9+10)
SlowQueries/slow-query-2-10     273ms ± 1%     236ms ± 1%  -13.59%  (p=0.000 n=10+10)
SlowQueries/slow-query-3-10    59.0ms ± 0%    59.1ms ± 1%     ~     (p=0.529 n=10+10)

name                         old alloc/op   new alloc/op   delta
SlowQueries/slow-query-1-10    11.9MB ± 0%    11.3MB ± 0%   -4.63%  (p=0.000 n=10+10)
SlowQueries/slow-query-2-10     154MB ± 0%      64MB ± 0%  -58.52%  (p=0.000 n=10+9)
SlowQueries/slow-query-3-10    47.5MB ± 0%    47.5MB ± 0%   -0.07%  (p=0.000 n=10+10)

name                         old allocs/op  new allocs/op  delta
SlowQueries/slow-query-1-10      110k ± 0%      106k ± 0%   -4.16%  (p=0.000 n=10+10)
SlowQueries/slow-query-2-10     1.14M ± 0%     0.45M ± 0%  -60.52%  (p=0.000 n=10+10)
SlowQueries/slow-query-3-10      359k ± 0%      358k ± 0%   -0.09%  (p=0.000 n=10+10)
```

DrewKimball committed:
opt: reduce allocations when building join multiplicity

This commit adds a `ComputeEquivClosureNoCopy` method to `FuncDepSet`
that performs the function of `ComputeEquivClosure` in-place on the given
`ColSet`. This allows callers to avoid an unnecessary copy if it is ok
to mutate the starting set.

This commit also modifies the multiplicity builder logic to simply build
the equivalence closure in one pass over the filters, rather than first
constructing a `FuncDepSet`. This significantly decreases allocations in
queries with many joins and equality filters.

Release note: None

DrewKimball commented:
I've just thought of one more improvement, so hold off on reviewing for a bit.

DrewKimball commented:
Updated benchmark:
```
name                         old time/op    new time/op    delta
SlowQueries/slow-query-1-10    19.3ms ± 0%    16.0ms ± 0%  -17.06%  (p=0.000 n=8+9)
SlowQueries/slow-query-2-10     273ms ± 0%     219ms ± 1%  -19.81%  (p=0.000 n=7+10)
SlowQueries/slow-query-3-10    59.1ms ± 0%    59.0ms ± 1%     ~     (p=0.315 n=10+10)

name                         old alloc/op   new alloc/op   delta
SlowQueries/slow-query-1-10    11.9MB ± 0%     8.1MB ± 0%  -31.92%  (p=0.000 n=10+10)
SlowQueries/slow-query-2-10     154MB ± 0%      48MB ± 0%  -68.81%  (p=0.000 n=10+6)
SlowQueries/slow-query-3-10    47.5MB ± 0%    47.3MB ± 0%   -0.49%  (p=0.000 n=8+10)

name                         old allocs/op  new allocs/op  delta
SlowQueries/slow-query-1-10      110k ± 0%       73k ± 0%  -33.49%  (p=0.000 n=10+10)
SlowQueries/slow-query-2-10     1.14M ± 0%     0.38M ± 0%  -66.52%  (p=0.000 n=10+10)
SlowQueries/slow-query-3-10      359k ± 0%      356k ± 0%   -0.88%  (p=0.000 n=9+10)
```

mgartner reviewed (approved):
Great finds and amazing results! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, 19 of 20 files at r2, 5 of 5 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-:-NBwBI5m1Bh5ryOqZD6j:bu252mh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw4jIt9XQ7ryv2uS6I):*
nit: I think you mean "exported", not "unexported".
___
*[`-- commits` line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw9LEXClQ9DSnBaPKT):*
Brilliant idea. So rather than build up an FD, then computing the equiv closure, each time `ComputeEquivClosureNoCopy` is called, the closure colset simply expands with new equivalent columns and the FD remains the same—is my understanding correct?
___
*[`pkg/sql/opt/norm/join_funcs.go` line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBwB0pI3Hq3MdtQqeeR:-NBwB0pI3Hq3MdtQqeeS:by5r8h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb05f0a052b62300ea9b5ca5d2ff371cfa6f8454/pkg/sql/opt/norm/join_funcs.go#L140)):*
> ```Go
> 	for i := range filters {
> 		fd := &filters[i].ScalarProps().FuncDeps
> 		filterEqCols := fd.ComputeEquivClosureNoCopy(fd.EquivReps())
> ```

nit: You could clarify here and below that `ComputeEquivClosureNoCopy` is safe because `fd.EquivReps` always returns a new `ColSet`.
___
*[`pkg/sql/opt/props/func_dep.go` line 713 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw8DHFDHAJB0yQiYfc:-NBw8DHFDHAJB0yQiYfd:b-vilvt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb05f0a052b62300ea9b5ca5d2ff371cfa6f8454/pkg/sql/opt/props/func_dep.go#L713)):*
> ```Go
> // closure in-place (e.g. the argument ColSet will be mutated). It should only
> // be used when it is ok to mutate the argument. This avoids allocations when
> // columns overflow the small set.
> ```

 nit: maybe clarify that "small set" refers to the small set of util.FastIntSet
___
*[`pkg/sql/opt/xform/optimizer.go` line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw3qCX8eT8uNZ8eyij:-NBw3qCX8eT8uNZ8eyik:b-umy6hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f162e2f5a3511d2dd78f3180ae701b026905dd0/pkg/sql/opt/xform/optimizer.go#L106)):*
> ```Go
> 	// scratchSort is used to avoid repeated allocations during sort enforcement.
> 	// It should be set to nil whenever the SortExpr is added to the memo, but
> 	// otherwise can be reused (this happens when adding the sort does not produce
> ```

nit: "It should be set to nil whenever the SortExpr is added to the memo **so that a new scratch SortExpr will be allocated the next time it is requested**, but otherwise can be reused ..."


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87920)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw4jIt9XQ7ryv2uS6I):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I think you mean "exported", not "unexported".
</blockquote></details>

Good catch, done.
___
*[`-- commits` line 45 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw9LEXClQ9DSnBaPKT):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Brilliant idea. So rather than build up an FD, then computing the equiv closure, each time `ComputeEquivClosureNoCopy` is called, the closure colset simply expands with new equivalent columns and the FD remains the same—is my understanding correct?
</blockquote></details>

Exactly. We get to avoid all the expensive FD-building that way.
___
*[`pkg/sql/opt/norm/join_funcs.go` line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBwB0pI3Hq3MdtQqeeR:-NBy8-w97cEZ3RGff9sn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb05f0a052b62300ea9b5ca5d2ff371cfa6f8454/pkg/sql/opt/norm/join_funcs.go#L140)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: You could clarify here and below that `ComputeEquivClosureNoCopy` is safe because `fd.EquivReps` always returns a new `ColSet`.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/props/func_dep.go` line 713 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw8DHFDHAJB0yQiYfc:-NBy8Cv6DraMh8iUkgnM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb05f0a052b62300ea9b5ca5d2ff371cfa6f8454/pkg/sql/opt/props/func_dep.go#L713)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

 nit: maybe clarify that "small set" refers to the small set of util.FastIntSet
</blockquote></details>

Done.
___
*[`pkg/sql/opt/xform/optimizer.go` line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-NBw3qCX8eT8uNZ8eyij:-NBy9LhYEwO5v41LISBk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f162e2f5a3511d2dd78f3180ae701b026905dd0/pkg/sql/opt/xform/optimizer.go#L106)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "It should be set to nil whenever the SortExpr is added to the memo **so that a new scratch SortExpr will be allocated the next time it is requested**, but otherwise can be reused ..."
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
@mgartner Thanks for taking a look! I discovered a bug in the sort-enforcer optimization - I wasn't accounting for the fact that `optimizeEnforcer` recurses into the input group member, which would then use the scratch sort. I changed the interface of `optimizeEnforcer` to pass a function that returns the enforcer in order to fix that, but it's definitely not as pretty.

Also, I had to update the `exprgen` code to handle the unexported field in `props.Relational`.

mgartner reviewed (approved):
> I changed the interface of optimizeEnforcer to pass a function that returns the enforcer in order to fix that, but it's definitely not as pretty.

Good catch. A little bit messier, but I couldn't think of a better way to solve it without a lot more lines of code. I think it's good for now.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 25 of 25 files at r6, 23 of 23 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87920#-:-NCM0f6f8zQ30qRQEP8I:b-t1csnq)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTR!

DrewKimball commented:
Test failure looks like a flake.

bors r+
